### ISSUE TITLE
Implement Adam optimizer

### ISSUE THREAD

**Issue by milancurcic** [modern-fortran/neural-fortran#115]
Proposed by @rweed in #114.

Paper: https://arxiv.org/abs/1412.6980

Currently, the optimizers module is only a stub and the only available optimizer (SGD) is hardcoded in the `network % train` method, with updating of weights progataing all the way down to individual concrete layer implementations. Some refactoring is needed to decouple the weight updates from concrete layer implementations and to allow defining optimizer algorithms in their own concrete types.

**Comment by rweed** [comment]
Milan

FYI, I found one Fortran implementation of Adam at
https://github.com/thchang/NN_MOD
Unfortunately, the comments in the code appear to suggest it was implemented
but never tested. Still looking for a batch normalization implementation (outside of Keras)
Also, one of my other "wants" , linear layers is trivial to implement. Took me all of about 5 minutes to do that in your existing code.

**Comment by milancurcic** [comment]
Thanks for the link to NN_MOD. I'd like to work on Adam first. I think it's easier to implement than batch norm, and it will drive the much needed refactor for optimizers in general (rather than them being hardcorded in the `network % train` subroutine).

Would you like to contribute the linear layer here as a PR? As I understand it, it's just a dense layer but without an activation. Are you just using a dense layer but with a "no-op" activation function (i.e. `y = x`)?

**Comment by rweed** [comment]
MIlan,

Thanks for working on this. I'm using Keras for now but I'm slowiy moving up to some very large training sets so it would be nice to be able to use neural fortran.

re. linear activation.

The linear activation function is just a passthrough (basically ReLU without the MAX test). The derivative is just 1.0

ie pure function linear(x) result(y)

        y = x
   end function
   pure function linear prime(x) result(dy)
       dy = 1.0
   end function

Most of the examples I've seen were folks are doing regression and not classification have used a linear output layer.

Re. batch normalization and dropout. When you feel you are up to doing one or the other I favor batch normalization over dropout. I use it first as a scaling pass prior to calling the first dense layer. I prefer this to doing a MINMAX scaling on the input data prior to training because it keeps the original unscaled data somewhat intact. Here is what I've implemented in Keras for the model to give you an idea of how I stack layers in a Sequential model.

model = Sequential()
if ibatchnorm > 0 :
  model.add(BatchNormalization())
for i in range (nhidden):
  model.add(Dense(nneurons, activation=actfun))
  if ibatchnorm > 0:
    model.add(BatchNormalization())
  if idropout > 0:
    model.add(Dropout(0.1))
model.add(Dense(1, activation='linear'))

What I'm trying to build is a surrogate for CFD calculations to predict loads on buildings. My training datasets will eventually approach 100K or more samples. Some initial tests I've run with 3 inputs (xyz position) of probe locations on the target buildings used around 3600 samples. So far batch normalization (in the sequence show above) using 4 layers with 500 neurons per layer and adam as the optimizer is giving the best results. Works a lot better than using just dropout layers. Also, as reported by other researchers, mixing dropout with batch normalization doesn't help (and can sometimes hurt).




________________________________________

**Comment by milancurcic** [comment]
@Spnetic-5, would you like to tackle this one next? I forgot whether you have a WIP implementation of Adam or AdaGrad?

**Comment by Spnetic-5** [comment]
Yes, Adam optimizer implementation is under progress; I'll make a PR soon.

**Comment by milancurcic** [comment]
Done by #150.
